Fire TransactionRemovedFromMempool callbacks from mempool

l2a5b1 opened the pull request:
This pull request fires TransactionRemovedFromMempool callbacks from the mempool and cleans up a bunch of code.

It also resolves the `txmempool -> validation -> validationinterface -> txmempool` circular dependency.

Ideally, `validationinterface` is a dumb component that doesn't have any knowledge of the sub-systems it sends its notifications to. The commit that aims to resolve this circular dependency by moving `txmempool` specific code out of `validationinterface` to `txmempool` where it belongs.

practicalswift commented on src/txmempool.cpp:
```diff
@@ -334,6 +335,21 @@ CTxMemPool::CTxMemPool(CBlockPolicyEstimator* estimator) :
+void CTxMemPool::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason reason)
+{
+    if (reason != MemPoolRemovalReason::BLOCK && reason != MemPoolRemovalReason::CONFLICT) {
+        GetMainSignals().TransactionRemovedFromMempool(ptx);
```
Could `ptx` be moved to avoid unnecessary copies?

251Labs replied:
Thanks @practicalswift. Good point. When `ptx` was passed by value I suspect we could have done that. However, per your suggestion https://github.com/bitcoin/bitcoin/pull/14384#discussion_r222600988 `ptx` is now passed by reference.

practicalswift commented on src/validationinterface.h:
```diff
@@ -174,13 +170,10 @@ class CMainSignals {
 
     void UpdatedBlockTip(const CBlockIndex *, const CBlockIndex *, bool fInitialDownload);
     void TransactionAddedToMempool(const CTransactionRef &);
+    void TransactionRemovedFromMempool(const CTransactionRef tx);
```
Should be named `ptx` to match definition :-)

251Labs replied:
Thanks @practicalswift fixed!

practicalswift commented on src/validationinterface.cpp:
```diff
@@ -148,6 +131,12 @@ void CMainSignals::TransactionAddedToMempool(const CTransactionRef &ptx) {
     });
 }
 
+void CMainSignals::TransactionRemovedFromMempool(const CTransactionRef ptx) {
```
Could be made `const` reference :-)

practicalswift commented:
Concept ACK

Let's get rid of these ugly circular dependencies! :-)

251Labs commented:
rebased fe6d863

thanks @practicalswift for the review - fe6d863 addresses your feedback. 

To answer your question https://github.com/bitcoin/bitcoin/pull/14384#discussion_r222600485: I don't think it is necessary to move the shared pointer as it is now passed by reference as per your suggestion: https://github.com/bitcoin/bitcoin/pull/14384#discussion_r222600988 .

251Labs commented:
Rebased 35a63b2. 

Updated the pull request description, since the `validation -> validationinterface -> validation` circular dependency was resolved in #16129 and removed from this pull request.

Travis build error seems unrelated.

l2a5b1 commented:
Rebased 80c3659

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -16,6 +16,12 @@
+// This map has to a separate global instead of a member of MainSignalsInstance,
+// because RegisterWithMempoolSignals is currently called before RegisterBackgroundSignalScheduler,
+// so MainSignalsInstance hasn't been created yet.
+static std::unordered_map<CTxMemPool*, boost::signals2::scoped_connection> g_connNotifyEntryRemoved;
```
Why is this not a CTxMemPool member?

MarcoFalke replied:
Actually can be removed in one go:

07c8420f4ddddde7103d8a564f918389357904c9

l2a5b1 replied:
Thanks @MarcoFalke! Really appreciate your feedback. Please see 04390b2.

l2a5b1 replied:
nit: Does it makes sense to remove ` CTxMemPool::EntryRemoved`'s method and move its implementation to `CTxMemPool::removeUnchecked` ?

```
void CTxMemPool::removeUnchecked(txiter it, MemPoolRemovalReason reason)
{
    NotifyEntryRemoved(ptx, reason);
    if (reason != MemPoolRemovalReason::BLOCK && reason != MemPoolRemovalReason::CONFLICT) {
        GetMainSignals().TransactionRemovedFromMempool(ptx);
    }
    ...
}
```

MarcoFalke replied:
Sure, if you wish so

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -333,6 +339,24 @@ CTxMemPool::CTxMemPool(CBlockPolicyEstimator* estimator) :
     nCheckFrequency = 0;
 }
 
+void CTxMemPool::RegisterSignals() {
```
Why can't this be called in the mempool constructor?

l2a5b1 commented:
04390b2 addresses feedback from @MarcoFalke.

MarcoFalke commented:
Adjusted title

MarcoFalke commented:
Concept ACK.

`TransactionRemovedFromMempool` is used to notify of removed tx, excluding txs that were already notified via `BlockConnected`. However, we will miss those notifications, if we forgot to call `RegisterWithMempoolSignals`. This happened in the unit tests (`setup_common`).

This pull fixes that and cleans up a bunch of code on the way.

l2a5b1 committed:
Fire TransactionRemovedFromMempool from mempool

This commit fires TransactionRemovedFromMempool callbacks from the
mempool and cleans up a bunch of code.

l2a5b1 commented:
28403ac addresses a nit (https://github.com/bitcoin/bitcoin/pull/14384#discussion_r306480219) and updates the commit message.

l2a5b1 commented:
Happy 1y anniversary! 🥳🎂

MarcoFalke commented:
Can we get this for 0.20, pls? Needs rebase

l2a5b1 commented:
@MarcoFalke sure, rebased 37fd92ba96224cfe11bacea069df02cead0c34e2

MarcoFalke commented:
ACK 37fd92ba96224cfe11bacea069df02cead0c34e2

0xB10C commented:
Concept ACK.

jnewbery commented:
Concept ACK. I've skimmed the code and it looks good. Will fully review and test next week.

I think a follow-up to this would be to remove the `pvtxConflicted` vector in `BlockConnected` and instead just call `TransactionRemovedFromMempool()` for conflicted transactions directly from `removeUnchecked()`. I don't see anywhere in the client code that requires the conflicted txs to be reported in the same notification as the block.

Doing that would allow us to remove the `NotifyEntryRemoved` signal from `CTxMempool`, the `conflictedTxs` from the `ConnectTrace` class, and some other code clean-up in validation.cpp.

jnewbery commented on src/validationinterface.cpp:
```diff
@@ -150,6 +127,12 @@ void CMainSignals::TransactionAddedToMempool(const CTransactionRef &ptx) {
     });
 }
 
+void CMainSignals::TransactionRemovedFromMempool(const CTransactionRef &ptx) {
```
Can you keep the `MemPoolRemovealReason reason` as an argument to this function please? #16688 adds validation interface logging, including logging the reason for the mempool removal.

I also think it makes sense for the client to be notified of the reason for the mempool reason.

jnewbery replied:
Can be done in a follow-up, or in #16688 when this gets merged.

jnewbery commented:
utACK 37fd92ba96224cfe11bacea069df02cead0c34e2

re https://github.com/bitcoin/bitcoin/pull/14384#discussion_r344524519: On second thoughts, there's no need to change `TransationRemovedFromMempool()` to take a `MemPoolRemovalReason` as an argument. If this PR gets merged before #16688 , then that PR should just add the argument there.

> I think a follow-up to this would be to remove the pvtxConflicted vector in BlockConnected 

I have a branch that does that here: https://github.com/jnewbery/bitcoin/tree/2019-11-remove-mempool-signals2. Will PR once this is merged.

l2a5b1 commented:
@jnewbery thanks for the review! 😃

ariard reviewed (approved):
Tested ACK 37fd92b

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/pull/14384/commits/37fd92ba96224cfe11bacea069df02cead0c34e2

This is basically some helpful code-shuffling; the change removes unneeded code to absorb the `TransactionRemovedFromMempool` signal into the existing `CValidationInterface` callback registration code, and avoids including mempool-specific header files into the more general validationinterface.

jonatack commented:
Concept ACK, kudos @l2a5b1 for staying with this. 

Built 37fd92ba96224cfe11bacea069df02cead0c34e2 rebased onto current master at 30521302f90e4856a7516867b32a4576fa6d98b3, tests pass, bitcoind running.

Now reviewing the code.

jonatack commented:
ACK 37fd92ba96224cfe11bacea069df02cead0c34e2

Nice cleanup that improves separation of concerns and enables removing `MempoolEntryRemoved`, `RegisterWithMempoolSignals`, and `UnregisterWithMempoolSignals`.

jonatack reviewed (commented):
ACK 37fd92ba96224cfe11bacea069df02cead0c34e2

Nice cleanup that improves separation of concerns and enables removing `MempoolEntryRemoved`, `RegisterWithMempoolSignals`, and `UnregisterWithMempoolSignals`.

TheBlueMatt commented:
This isn't strictly worse, but its still super awkward....CTxMemPool is (mostly) a "dumb datastructure", with all the adding/removing happening in validation.cpp. Currently its double awkward cause they're signals that go back to validation so that validation can fire appropriate events, but just firing the events inside a datastructure is also really awkward. Is there an equally-easy way to move the event-firing back into validation.cpp?

jnewbery commented:
> Is there an equally-easy way to move the event-firing back into validation.cpp?

I don't think there is. That would require that every site that validation calls a `CTxMemPool` method that could result in transactions being removed would need to be passed back a vector of transactions that it should fire the events for. A quick search shows at least the following sites would need to be updated:

- validation's `UpdateMempoolForReorg()` calling into `mempool.removeRecursive()`
- validation's `LimitMempoolSize()` calling into `pool.Expire()`
- validation's `LimitMempoolSize()` calling into `pool.TrimToSize()`

There are probably more. I don't think those call sites have *ever* handled firing the events, and adding that functionality would be new data structures, function signatures and logic.

This seems to be a strict improvement to me. The events are being fired from exactly the same place as before, except that there's no weird indirection in the validation interface.

MarcoFalke commented:
Re-run ci

TheBlueMatt commented:
Right, certainly don't interpret my comment as a NACK, then. Longer-term it would be nice, indeed, to move the event-firing into those callsites, cause the responsibility right now is very awkwardly split (hence why we're checking the reason in the first place - validation fires some similar things), but this isn't worse than master and if other PR's need this, do it!

laanwj commented:
argh, needs trivial rebase for the linter `test/lint/lint-circular-dependencies.sh `

l2a5b1 commented:
> argh

@laanwj lol, no worries, rebased e20c72f9f076681def325b5b5fa53bccda2b0eab

jnewbery commented:
ACK e20c72f9f076681def325b5b5fa53bccda2b0eab

ariard commented:
ACK e20c72f, only rebased.

jonatack commented:
Posthumous ACK e20c72f9f076681def325b5b5fa53bccda2b0eab
